+ 2 feature selection methods were implemented in the paper; removing features with low variance and removing colinear variables.Both will be used here.

In [1]:
import pandas as pd 


In [2]:
df = pd.read_csv("ache_pchem.csv")

In [3]:
df.head()

Label  PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP4  \
0      0           1           1           1           0           0   
1      0           1           1           0           0           0   
2      0           1           1           1           1           0   
3      0           1           1           1           1           0   
4      0           1           1           1           1           0   

   PubchemFP5  PubchemFP6  PubchemFP7  PubchemFP8  ...  PubchemFP871  \
0           0           0           0           0  ...             0   
1           0           0           0           0  ...             0   
2           0           0           0           0  ...             0   
3           0           0           0           0  ...             0   
4           0           0           0           0  ...             0   

   PubchemFP872  PubchemFP873  PubchemFP874  PubchemFP875  PubchemFP876  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   PubchemFP877  PubchemFP878  PubchemFP879  PubchemFP880  
0             0             0             0             0  
1             0             0             0             0  
2             0             0             0             0  
3             0             0             0             0  
4             0             0             0             0  

[5 rows x 882 columns]

In [4]:
X= df.drop(["Label"],axis=1)

In [5]:
from sklearn.feature_selection import VarianceThreshold

In [6]:
var = VarianceThreshold(0)

In [7]:
var.fit(X)

VarianceThreshold(threshold=0)

In [8]:
var.get_support()

array([ True,  True,  True,  True, False, False, False, False, False,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True, False,
       False,  True, False,  True,  True, False,  True,  True, False,
       False,  True,  True,  True, False, False, False,  True,  True,
       False,  True,  True, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True,  True,
       False,  True, False, False, False,  True,  True, False, False,
       False, False,

In [11]:
X = pd.DataFrame(var.transform(X),columns=X.columns[var.get_support()])

In [12]:
X.head()

PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP9  PubchemFP10  \
0           1           1           1           0           1            1   
1           1           1           0           0           1            1   
2           1           1           1           1           1            1   
3           1           1           1           1           1            1   
4           1           1           1           1           1            1   

   PubchemFP11  PubchemFP12  PubchemFP13  PubchemFP14  ...  PubchemFP854  \
0            1            0            0            1  ...             0   
1            1            0            0            1  ...             0   
2            1            1            1            1  ...             0   
3            1            1            1            1  ...             0   
4            1            1            1            1  ...             0   

   PubchemFP860  PubchemFP861  PubchemFP862  PubchemFP863  PubchemFP864  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   PubchemFP866  PubchemFP868  PubchemFP872  PubchemFP875  
0             0             0             0             0  
1             0             0             0             0  
2             0             0             0             0  
3             0             0             0             0  
4             0             0             0             0  

[5 rows x 639 columns]

In [13]:
df = pd.concat([X,df["Label"]],axis=1)

In [14]:
df.head()

PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP9  PubchemFP10  \
0           1           1           1           0           1            1   
1           1           1           0           0           1            1   
2           1           1           1           1           1            1   
3           1           1           1           1           1            1   
4           1           1           1           1           1            1   

   PubchemFP11  PubchemFP12  PubchemFP13  PubchemFP14  ...  PubchemFP860  \
0            1            0            0            1  ...             0   
1            1            0            0            1  ...             0   
2            1            1            1            1  ...             0   
3            1            1            1            1  ...             0   
4            1            1            1            1  ...             0   

   PubchemFP861  PubchemFP862  PubchemFP863  PubchemFP864  PubchemFP866  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   PubchemFP868  PubchemFP872  PubchemFP875  Label  
0             0             0             0      0  
1             0             0             0      0  
2             0             0             0      0  
3             0             0             0      0  
4             0             0             0      0  

[5 rows x 640 columns]

In [15]:
df_sub = pd.read_csv("ache_sub.csv")

In [16]:
X_ = df_sub.drop(["Label"],axis=1)

In [17]:
X_.head()

SubFP1  SubFP2  SubFP3  SubFP4  SubFP5  SubFP6  SubFP7  SubFP8  SubFP9  \
0       1       1       0       1       0       0       0       0       0   
1       0       1       0       0       0       0       0       0       0   
2       0       1       0       0       0       0       0       0       0   
3       0       1       0       0       0       0       0       0       0   
4       0       1       0       0       0       0       0       0       0   

   SubFP10  ...  SubFP298  SubFP299  SubFP300  SubFP301  SubFP302  SubFP303  \
0        0  ...         0         0         1         1         1         0   
1        0  ...         0         0         1         1         0         0   
2        0  ...         0         0         1         1         1         0   
3        0  ...         0         0         1         1         1         0   
4        0  ...         0         0         1         1         1         0   

   SubFP304  SubFP305  SubFP306  SubFP307  
0         0         0         0         1  
1         0         0         0         1  
2         0         0         0         1  
3         0         0         0         1  
4         0         0         0         1  

[5 rows x 307 columns]

In [18]:
v = VarianceThreshold(0)

In [19]:
X_ = pd.DataFrame(v.fit_transform(X_),columns=X_.columns[v.get_support()])

In [20]:
X_.head()

SubFP1  SubFP2  SubFP3  SubFP4  SubFP5  SubFP6  SubFP8  SubFP9  SubFP10  \
0       1       1       0       1       0       0       0       0        0   
1       0       1       0       0       0       0       0       0        0   
2       0       1       0       0       0       0       0       0        0   
3       0       1       0       0       0       0       0       0        0   
4       0       1       0       0       0       0       0       0        0   

   SubFP11  ...  SubFP294  SubFP295  SubFP296  SubFP297  SubFP298  SubFP299  \
0        0  ...         0         1         0         0         0         0   
1        0  ...         0         1         0         0         0         0   
2        0  ...         0         1         0         0         0         0   
3        0  ...         0         1         0         0         0         0   
4        0  ...         0         1         0         0         0         0   

   SubFP300  SubFP301  SubFP302  SubFP303  
0         1         1         1         0  
1         1         1         0         0  
2         1         1         1         0  
3         1         1         1         0  
4         1         1         1         0  

[5 rows x 147 columns]

In [22]:
df_sub = pd.concat([X_,df_sub["Label"]],axis=1)

In [23]:
df_sub

SubFP1  SubFP2  SubFP3  SubFP4  SubFP5  SubFP6  SubFP8  SubFP9  SubFP10  \
0          1       1       0       1       0       0       0       0        0   
1          0       1       0       0       0       0       0       0        0   
2          0       1       0       0       0       0       0       0        0   
3          0       1       0       0       0       0       0       0        0   
4          0       1       0       0       0       0       0       0        0   
...      ...     ...     ...     ...     ...     ...     ...     ...      ...   
4999       0       1       0       0       0       0       0       0        0   
5000       0       1       0       0       0       0       0       0        0   
5001       0       1       0       0       0       0       0       0        0   
5002       0       1       0       0       0       0       0       0        0   
5003       0       1       0       0       0       0       0       0        0   

      SubFP11  ...  SubFP295  SubFP296  SubFP297  SubFP298  SubFP299  \
0           0  ...         1         0         0         0         0   
1           0  ...         1         0         0         0         0   
2           0  ...         1         0         0         0         0   
3           0  ...         1         0         0         0         0   
4           0  ...         1         0         0         0         0   
...       ...  ...       ...       ...       ...       ...       ...   
4999        0  ...         1         0         0         0         0   
5000        0  ...         1         0         0         0         0   
5001        0  ...         1         0         0         0         0   
5002        0  ...         1         0         0         0         0   
5003        0  ...         1         0         0         0         0   

      SubFP300  SubFP301  SubFP302  SubFP303  Label  
0            1         1         1         0      0  
1            1         1         0         0      0  
2            1         1         1         0      0  
3            1         1         1         0      0  
4            1         1         1         0      0  
...        ...       ...       ...       ...    ...  
4999         1         1         1         0      1  
5000         1         1         1         0      1  
5001         1         1         1         0      1  
5002         1         1         1         0      1  
5003         1         1         1         0      1  

[5004 rows x 148 columns]

+ **Removing Multicolinearity**

Collinearity is a very common problem in machine learning projects. It is the correlation between the features of a dataset and it can reduce the performance of our models because it increases variance and the number of dimensions. It becomes worst when you have to work with unsupervised models.

In [25]:
#!pip install collinearity

In [26]:
from collinearity import SelectNonCollinear

In [27]:
df.head()

PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP9  PubchemFP10  \
0           1           1           1           0           1            1   
1           1           1           0           0           1            1   
2           1           1           1           1           1            1   
3           1           1           1           1           1            1   
4           1           1           1           1           1            1   

   PubchemFP11  PubchemFP12  PubchemFP13  PubchemFP14  ...  PubchemFP860  \
0            1            0            0            1  ...             0   
1            1            0            0            1  ...             0   
2            1            1            1            1  ...             0   
3            1            1            1            1  ...             0   
4            1            1            1            1  ...             0   

   PubchemFP861  PubchemFP862  PubchemFP863  PubchemFP864  PubchemFP866  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   PubchemFP868  PubchemFP872  PubchemFP875  Label  
0             0             0             0      0  
1             0             0             0      0  
2             0             0             0      0  
3             0             0             0      0  
4             0             0             0      0  

[5 rows x 640 columns]

In [31]:
selector = SelectNonCollinear(0.7)

In [32]:
X.head()

PubchemFP0  PubchemFP1  PubchemFP2  PubchemFP3  PubchemFP9  PubchemFP10  \
0           1           1           1           0           1            1   
1           1           1           0           0           1            1   
2           1           1           1           1           1            1   
3           1           1           1           1           1            1   
4           1           1           1           1           1            1   

   PubchemFP11  PubchemFP12  PubchemFP13  PubchemFP14  ...  PubchemFP854  \
0            1            0            0            1  ...             0   
1            1            0            0            1  ...             0   
2            1            1            1            1  ...             0   
3            1            1            1            1  ...             0   
4            1            1            1            1  ...             0   

   PubchemFP860  PubchemFP861  PubchemFP862  PubchemFP863  PubchemFP864  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   PubchemFP866  PubchemFP868  PubchemFP872  PubchemFP875  
0             0             0             0             0  
1             0             0             0             0  
2             0             0             0             0  
3             0             0             0             0  
4             0             0             0             0  

[5 rows x 639 columns]

In [33]:
y = df["Label"]

In [42]:
selector.fit(X.values,y)

In [44]:
mask = selector.get_support()

In [49]:
df_pc = pd.DataFrame(selector.transform(X.values),columns=X.columns[selector.get_support()])

In [50]:
df_pc.head()

PubchemFP0  PubchemFP3  PubchemFP9  PubchemFP11  PubchemFP12  PubchemFP13  \
0           1           0           1            1            0            0   
1           1           0           1            1            0            0   
2           1           1           1            1            1            1   
3           1           1           1            1            1            1   
4           1           1           1            1            1            1   

   PubchemFP15  PubchemFP16  PubchemFP17  PubchemFP19  ...  PubchemFP841  \
0            1            0            0            1  ...             0   
1            1            0            0            0  ...             0   
2            1            1            0            1  ...             0   
3            1            1            0            1  ...             0   
4            1            1            0            1  ...             0   

   PubchemFP842  PubchemFP843  PubchemFP845  PubchemFP847  PubchemFP860  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   PubchemFP861  PubchemFP866  PubchemFP868  PubchemFP875  
0             0             0             0             0  
1             0             0             0             0  
2             0             0             0             0  
3             0             0             0             0  
4             0             0             0             0  

[5 rows x 269 columns]

In [51]:
df_pc["Label"] = df["Label"]

In [52]:
df_pc

PubchemFP0  PubchemFP3  PubchemFP9  PubchemFP11  PubchemFP12  \
0              1           0           1            1            0   
1              1           0           1            1            0   
2              1           1           1            1            1   
3              1           1           1            1            1   
4              1           1           1            1            1   
...          ...         ...         ...          ...          ...   
4999           0           0           1            1            1   
5000           0           0           1            1            1   
5001           0           0           1            1            1   
5002           0           0           1            1            1   
5003           0           0           1            1            1   

      PubchemFP13  PubchemFP15  PubchemFP16  PubchemFP17  PubchemFP19  ...  \
0               0            1            0            0            1  ...   
1               0            1            0            0            0  ...   
2               1            1            1            0            1  ...   
3               1            1            1            0            1  ...   
4               1            1            1            0            1  ...   
...           ...          ...          ...          ...          ...  ...   
4999            0            1            1            0            0  ...   
5000            0            1            1            0            0  ...   
5001            0            1            1            0            0  ...   
5002            0            1            1            0            0  ...   
5003            1            1            1            0            1  ...   

      PubchemFP842  PubchemFP843  PubchemFP845  PubchemFP847  PubchemFP860  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
4999             0             0             0             0             0   
5000             0             0             0             0             0   
5001             0             0             0             0             0   
5002             0             0             0             0             0   
5003             0             0             0             0             0   

      PubchemFP861  PubchemFP866  PubchemFP868  PubchemFP875  Label  
0                0             0             0             0      0  
1                0             0             0             0      0  
2                0             0             0             0      0  
3                0             0             0             0      0  
4                0             0             0             0      0  
...            ...           ...           ...           ...    ...  
4999             0             0             0             0      1  
5000             0             0             0             0      1  
5001             0             0             0             0      1  
5002             0             0             0             0      1  
5003             0             0             0             0      1  

[5004 rows x 270 columns]

In [53]:
df_pc.to_csv("pubchem_dataset.csv",sep="\t",index=None)

In [54]:
df_sub.head()

SubFP1  SubFP2  SubFP3  SubFP4  SubFP5  SubFP6  SubFP8  SubFP9  SubFP10  \
0       1       1       0       1       0       0       0       0        0   
1       0       1       0       0       0       0       0       0        0   
2       0       1       0       0       0       0       0       0        0   
3       0       1       0       0       0       0       0       0        0   
4       0       1       0       0       0       0       0       0        0   

   SubFP11  ...  SubFP295  SubFP296  SubFP297  SubFP298  SubFP299  SubFP300  \
0        0  ...         1         0         0         0         0         1   
1        0  ...         1         0         0         0         0         1   
2        0  ...         1         0         0         0         0         1   
3        0  ...         1         0         0         0         0         1   
4        0  ...         1         0         0         0         0         1   

   SubFP301  SubFP302  SubFP303  Label  
0         1         1         0      0  
1         1         0         0      0  
2         1         1         0      0  
3         1         1         0      0  
4         1         1         0      0  

[5 rows x 148 columns]

In [58]:
sele = SelectNonCollinear(0.7)

In [59]:
sele.fit(X_.values,y)

In [60]:
df_sub = pd.DataFrame(sele.transform(X_.values),columns=X_.columns[sele.get_support()])

In [61]:
df_sub.head()

SubFP1  SubFP2  SubFP3  SubFP4  SubFP5  SubFP6  SubFP8  SubFP9  SubFP10  \
0       1       1       0       1       0       0       0       0        0   
1       0       1       0       0       0       0       0       0        0   
2       0       1       0       0       0       0       0       0        0   
3       0       1       0       0       0       0       0       0        0   
4       0       1       0       0       0       0       0       0        0   

   SubFP11  ...  SubFP288  SubFP291  SubFP292  SubFP293  SubFP295  SubFP296  \
0        0  ...         0         0         0         0         1         0   
1        0  ...         0         0         0         0         1         0   
2        0  ...         0         0         0         0         1         0   
3        0  ...         0         0         0         0         1         0   
4        0  ...         0         0         0         0         1         0   

   SubFP297  SubFP300  SubFP302  SubFP303  
0         0         1         1         0  
1         0         1         0         0  
2         0         1         1         0  
3         0         1         1         0  
4         0         1         1         0  

[5 rows x 125 columns]

In [62]:
df_sub["Label"] = df["Label"]

In [63]:
df_sub.head()

SubFP1  SubFP2  SubFP3  SubFP4  SubFP5  SubFP6  SubFP8  SubFP9  SubFP10  \
0       1       1       0       1       0       0       0       0        0   
1       0       1       0       0       0       0       0       0        0   
2       0       1       0       0       0       0       0       0        0   
3       0       1       0       0       0       0       0       0        0   
4       0       1       0       0       0       0       0       0        0   

   SubFP11  ...  SubFP291  SubFP292  SubFP293  SubFP295  SubFP296  SubFP297  \
0        0  ...         0         0         0         1         0         0   
1        0  ...         0         0         0         1         0         0   
2        0  ...         0         0         0         1         0         0   
3        0  ...         0         0         0         1         0         0   
4        0  ...         0         0         0         1         0         0   

   SubFP300  SubFP302  SubFP303  Label  
0         1         1         0      0  
1         1         0         0      0  
2         1         1         0      0  
3         1         1         0      0  
4         1         1         0      0  

[5 rows x 126 columns]

In [64]:
df_sub.to_csv("subfp_dataset.csv",sep="\t",index=None)

+ A total of 269 and 125 fingerprints results from feature selection step.